runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcDirectGetWorkStream.java

```diff
@@ -0,0 +1,328 @@
+                    .setMaxItems(adjustment.items())
+                    .setMaxBytes(adjustment.bytes()))
+            .build();
+
+    executor()
+        .execute(
+            () -> {
+              try {
+                send(extension);
+              } catch (IllegalStateException e) {
+                // Stream was closed.
+              }
+            });
+  }
+
+  @Override
+  protected synchronized void onNewStream() {
+    workItemBuffers.clear();
+    // Add the current inflight budget to the next adjustment. Only positive values are allowed here
+    // with negatives defaulting to 0, since GetWorkBudgets cannot be created with negative values.
+    GetWorkBudget budgetAdjustment = nextBudgetAdjustment.get().apply(inflightBudget.get());
+    inflightBudget.set(budgetAdjustment);
+    send(
+        StreamingGetWorkRequest.newBuilder()
+            .setRequest(
+                request
+                    .toBuilder()
+                    .setMaxBytes(budgetAdjustment.bytes())
+                    .setMaxItems(budgetAdjustment.items()))
+            .build());
+
+    // We just sent the budget, reset it.
+    nextBudgetAdjustment.set(GetWorkBudget.noBudget());
+  }
+
+  @Override
+  protected boolean hasPendingRequests() {
+    return false;
+  }
+
+  @Override
+  public void appendSpecificHtml(PrintWriter writer) {
+    // Number of buffers is same as distinct workers that sent work on this stream.
+    writer.format(
+        "GetWorkStream: %d buffers, %s inflight budget allowed.",
+        workItemBuffers.size(), inflightBudget.get());
+  }
+
+  @Override
+  public void sendHealthCheck() {
+    send(HEALTH_CHECK_REQUEST);
+  }
+
+  @Override
+  protected void onResponse(StreamingGetWorkResponseChunk chunk) {
+    getWorkThrottleTimer.stop();
+    WorkItemBuffer workItemBuffer =
+        workItemBuffers.computeIfAbsent(chunk.getStreamId(), unused -> new WorkItemBuffer());
+    workItemBuffer.append(chunk);
+
+    // The entire WorkItem has been received, it is ready to be processed.
+    if (chunk.getRemainingBytesForWorkItem() == 0) {
+      workItemBuffer.runAndReset();
+      // Record the fact that there are now fewer outstanding messages and bytes on the stream.
+      inflightBudget.updateAndGet(budget -> budget.subtract(1, workItemBuffer.bufferedSize()));
+    }
+  }
+
+  @Override
+  protected void startThrottleTimer() {
+    getWorkThrottleTimer.start();
+  }
+
+  @Override
+  public synchronized void adjustBudget(long itemsDelta, long bytesDelta) {
+    nextBudgetAdjustment.set(nextBudgetAdjustment.get().apply(itemsDelta, bytesDelta));
+    sendRequestExtension();
+  }
+
+  @Override
+  public GetWorkBudget remainingBudget() {
+    // Snapshot the current budgets.
+    GetWorkBudget currentPendingResponseBudget = pendingResponseBudget.get();
+    GetWorkBudget currentNextBudgetAdjustment = nextBudgetAdjustment.get();
+    GetWorkBudget currentInflightBudget = inflightBudget.get();
+
+    return GetWorkBudget.builder()
```

Comment:
Think you could simplify as
currentPendingResponseBudget.apply(currentNextBudgetAdjustment).apply(currentInflightBudget)

Reply:
Done